### PR TITLE
Add user metadata

### PR DESCRIPTION
**Pull request by maeldebon**
🔗 **Related PRs:**
- [substra](https://github.com/SubstraFoundation/substra/pull/143)
- [substra-chaincode](https://github.com/SubstraFoundation/substra-chaincode/pull/111)
- [substra-backend](https://github.com/SubstraFoundation/substra-backend/pull/186)
- [substra-frontend (coming soon)](#)

The goal of these PRs is to add a `metadata` field in every asset in order to be able to store information related to this asset. It could be, for instance, a creation_ts, an epoch, a project_name, ...

Update 04/24:
> The following assets are missing in the tests:
> - dataset,
> - aggregatetuple (WIP),
> - compute_plan.

> The issue I currently have is that the metadata field returned by the aggregatetuple tests is set to an empty dict, but I pass:
{"foo": "bar"}
> I need to find a way to get the right output.

Update 06/03:
> The compute_plan asset cannot have a metadata field for now. It could be done later.

### CODE REVIEW

File: substratest/assets.py [hunk 1]
```diff
@@ -194,13 +194,15 @@ class DataSample(_Asset):
 class ObjectiveDataset(_InternalStruct):
     dataset_key: str
     data_sample_keys: typing.List[str]
+    metadata: typing.Dict
```
Comment (jmorel) [inline substratest/assets.py]: The ObjectiveDataset isn't really an asset, it's a part of the Objective. It won't have metadata attached to it.

File: substratest/assets.py [hunk 2]
```diff
@@ -194,13 +194,15 @@ class DataSample(_Asset):
...
 
 
 class Dataset(_Asset):
     key: str
     name: str
     owner: str
     objective_key: str
+    metadata: typing.Dict
```
Comment (jmorel) [inline substratest/assets.py]: Since you specified in the backend that metadata is a dict of strings, you could also specify it here with the syntax typing.Dict[str, str] (keys are str, values are str)

Comment (jmorel) [reply substratest/assets.py]: This would also apply to other assets.

File: substratest/assets.py [hunk 3]
```diff
@@ -240,6 +244,7 @@ class TesttupleDataset(_InternalStruct):
     perf: float
     keys: typing.List[str]
     worker: str
+    metadata: typing.Dict
```
Comment (jmorel) [inline substratest/assets.py]: As for the ObjectiveDataset, this isn't an Asset and won't have metadata.

File: substratest/factory.py [hunk 4]
```diff
@@ -184,6 +184,7 @@ class DatasetSpec(_Spec):
     data_opener: str
     type: str
     description: str
+    metadata: typing.Dict
```
Comment (jmorel) [inline substratest/factory.py]: As for assets.py, you could be more precise here and use `typing.Dict[str,str]`

File: substratest/factory.py [hunk 5]
```diff
@@ -346,21 +357,24 @@ class _BaseComputePlanSpec(_Spec, abc.ABC):
     aggregatetuples: typing.List[ComputePlanAggregatetupleSpec]
     testtuples: typing.List[ComputePlanTesttupleSpec]
 
-    def add_traintuple(self, algo, dataset, data_samples, in_models=None, tag=''):
+    def add_traintuple(self, algo, dataset, data_samples, in_models=None, tag='', metadata=None):
         in_models = in_models or []
+        metadata = metadata or {}
```
Comment (jmorel) [inline substratest/factory.py]: This line does the same as L370, you'll need to use one or the other!

Comment (jmorel) [reply substratest/factory.py]: This also applies to add_aggregatuple and add_composite_traintuple.

Comment (jmorel) [reply substratest/factory.py]: I suggest you use everywhere the pattern you're already using for add_testtuple.

File: substratest/factory.py [hunk 6]
```diff
@@ -458,7 +476,8 @@ def create_data_sample(self, content=None, datasets=None, test_only=False):
             data_manager_keys=[d.key for d in datasets],
         )
 
-    def create_dataset(self, objective=None, permissions=None):
+    def create_dataset(self, objective=None, permissions=None, metadata=None):
+        metadata = metadata or {}
```
Comment (jmorel) [inline substratest/factory.py]: This doubles with L501.

File: substratest/factory.py [hunk 7]
```diff
@@ -479,12 +498,14 @@ def create_dataset(self, objective=None, permissions=None):
...
             type='Test',
+            metadata=metadata or {},
             description=str(description_path),
             objective_key=objective.key if objective else None,
             permissions=permissions or DEFAULT_PERMISSIONS,
         )
 
-    def create_objective(self, dataset=None, data_samples=None, permissions=None):
+    def create_objective(self, dataset=None, data_samples=None, permissions=None, metadata=None):
+        metadata = metadata or {}
```
Comment (jmorel) [inline substratest/factory.py]: The pattern you're using here is different from the one you used for compute plan testtuples. You should chose one and apply it throughout the file.

File: substratest/assets.py (hunk 2 above)
Comment (maeldebon) [reply substratest/assets.py]: Yes, indeed!

File: substratest/factory.py (hunk 5 above)
Comment (maeldebon) [reply substratest/factory.py]: You're right, I forgot to remove it :/

File: substratest/assets.py [hunk 8]
```diff
@@ -331,6 +337,7 @@ class Testtuple(_Asset, _FutureMixin):
     certified: bool
     rank: int
     tag: str
+    metadata: typing.Dict[str, str] = None
```
Comment (jmorel) [inline substratest/assets.py:340]: The main description says testtuples are tested, but there's still `= None` here. Could you remove it?

File: substratest/factory.py [hunk 9]
```diff
@@ -371,13 +383,14 @@ def add_aggregatetuple(self, aggregate_algo, worker, in_models=None, tag=''):
             worker=worker,
             in_models_ids=[t.id for t in in_models],
             tag=tag,
+            metadata=metadata or {}
```
Comment (jmorel) [inline substratest/factory.py:386]: In half the cases, you've been doing `metadata=metadata` and in the other half `metadata=metadata or {}`. You should chose one of the two solutions and stick with it. When you do, could you also explain why you chose one over the other?

File: tests/test_network.py [hunk 10]
```diff
@@ -17,10 +17,20 @@ def test_connection_to_nodes(clients):
...
 
+#
+# def test_add_dataset_with_metadata(factory, client):
+#     spec = factory.create_dataset(metadata={'foo': 'bar'})
+#     dataset = client.add_dataset(spec)
+#     assert dataset.metadata == {'foo': 'bar'}
+#
+#     dataset_copy = client.get_dataset(dataset.key)
+#     assert dataset == dataset_copy
+
```
Comment (jmorel) [inline tests/test_network.py:33]: Could you cleanup these comments?

File: tests/test_network.py [hunk 11]
```diff
@@ -124,6 +134,159 @@ def test_add_objective(factory, client):
...
+    too_long = 'foo' * 40
+    # raise an error if the key is too long (more than 50 chars)
+    spec = method_create(metadata={too_long: "bar"})
+    with pytest.raises(substra.exceptions.InvalidRequest):
+        method_add(spec)
+
+    # raise an error if the value is too long (more than 100 chars)
+    spec = method_create(metadata={"foo": too_long})
+    with pytest.raises(substra.exceptions.InvalidRequest):
+        method_add(spec)
```
Comment (jmorel) [inline tests/test_network.py:183]: I feel like all the errors should be handled in the same way: always raise an InvalidRequest. Do you think it would be possible to update the backend and sdk code so that we have a consistent behavior from the user standpoint?

File: tests/test_network.py [hunk 12]
```diff
@@ -124,6 +134,159 @@ def test_add_objective(factory, client):
...
+
+def test_metadata_aggregatetuple(factory, client, default_dataset):
+    composite_algo_spec = factory.create_composite_algo()
+    composite_algo = client.add_composite_algo(composite_algo_spec)
+    composite_traintuple_spec = factory.create_composite_traintuple(
+        algo=composite_algo,
+        dataset=default_dataset,
+        data_samples=default_dataset.train_data_sample_keys,
+    )
+    composite_traintuple = client.add_composite_traintuple(composite_traintuple_spec)
```
Comment (jmorel) [inline tests/test_network.py:272]: You can create an aggregatetuple without inmodels (see the discussion at https://github.com/SubstraFoundation/substra-backend/pull/189). This would simplify this test a great deal.

File: tests/test_network.py [hunk 13]
```diff
@@ -124,6 +134,159 @@ def test_add_objective(factory, client):
...
+    'aggregate_algo',
+    'composite_algo',
+])
+def test_metadata(factory, client, asset_name):
+    method_create = getattr(factory, f"create_{asset_name}")
+    method_add = getattr(client, f"add_{asset_name}")
+
+    # add an asset with metadata
+    spec = method_create(metadata={"foo": "bar"})
+    asset = method_add(spec)
```
Comment (jmorel) [inline tests/test_network.py:149]: Since this test is trying out multiple metadata values, we could parametrize the metadata as well.

There could be another layer of parametrization to specify what the input metadata value, what the output should and which error should be raised.

Here is an example of combining parametrization:

```python
@pytest.mark.parametrize('y', (1, 2))
@pytest.mark.parametrize('x', ('a', 'b'))
def test_foo(x, y):
    assert True
```
This will be executed in the following order:
- a-1
- a-2
- b-1
- b-2

See also for how to handle exceptions https://github.com/SubstraFoundation/substra-backend/pull/189

**Comment by jmorel** [comment]
When you'll be testing it for compute plans, don't forget that you'll also need to write tests for all compute plan assets.

File: tests/test_network.py [hunk 14]
```diff
@@ -124,6 +134,159 @@ def test_add_objective(factory, client):
...
+    spec = method_create(metadata={"foo": too_long})
+    with pytest.raises(substra.exceptions.InvalidRequest):
+        method_add(spec)
+
+
+@pytest.mark.parametrize('asset_name, algo_type', [
+    ('traintuple', 'algo'),
+    ('composite_traintuple', 'composite_algo'),
+])
+def test_metadata_traintuple(factory, client, asset_name, algo_type, default_dataset):
```
Comment (jmorel) [inline tests/test_network.py:190]: This test creates a lot of tuples. These tuples should be canceled if possible and the test should not end until the tuples are either done or canceled. Otherwise the execution of tuples in other tests might be delayed by these ones and timeout.

File: substratest/factory.py [hunk 15]
```diff
@@ -179,6 +179,7 @@ class DatasetSpec(_Spec):
     data_opener: str
     type: str
     description: str
+    metadata: typing.Dict = None
```
Comment (jmorel) [inline substratest/factory.py:182]: You could even go further and specify `typing.Dict[str, str] = None`, but this would also require to change the test on `metadata={'foo': {'bar': 'baz'}}` in the same way you did for `metadata='foo'`

File: substratest/factory.py [hunk 16]
```diff
@@ -379,15 +392,17 @@ def add_composite_traintuple(self, composite_algo, dataset=None, data_samples=No
             in_trunk_model_id=in_trunk_model.id if in_trunk_model else None,
             out_trunk_model_permissions=out_trunk_model_permissions or DEFAULT_OUT_MODEL_PERMISSIONS,
             tag=tag,
+            metadata=metadata or {}
```
Comment (jmorel) [inline substratest/factory.py:395]: Same as above.

File: substratest/factory.py [hunk 17]
```diff
@@ -379,15 +392,17 @@ def add_composite_traintuple(self, composite_algo, dataset=None, data_samples=No
...
         self.composite_traintuples.append(spec)
         return spec
 
-    def add_testtuple(self, objective, traintuple_spec, tag=None):
+    def add_testtuple(self, objective, traintuple_spec, tag=None, metadata=None):
         spec = ComputePlanTesttupleSpec(
             objective_key=objective.key,
             traintuple_id=traintuple_spec.id,
             tag=tag or '',
+            metadata=metadata or {}
```
Comment (jmorel) [inline substratest/factory.py:405]: Same as above.

File: tests/test_network.py [hunk 18]
```diff
@@ -17,10 +17,20 @@ def test_connection_to_nodes(clients):
 def test_add_dataset(factory, client):
     spec = factory.create_dataset()
     dataset = client.add_dataset(spec)
+    # assert dataset.metadata == {}
```
Comment (jmorel) [inline tests/test_network.py:20]: I feel this shouldn't be commented at all.

File: tests/test_network.py [hunk 19]
```diff
@@ -124,6 +134,162 @@ def test_add_objective(factory, client):
...
+    spec = factory.create_aggregatetuple(
+        algo=aggregate_algo,
+        worker='MyOrg1MSP',
+        traintuples=[composite_traintuple],
+        metadata={"foo": "bar"}
+    )
+    asset = client.add_aggregatetuple(spec)
+
+    print(asset)
+    print(composite_traintuple)
```
Comment (jmorel) [inline tests/test_network.py:288]: These should be removed.

File: tests/test_network.py [hunk 20]
```diff
@@ -124,6 +134,162 @@ def test_add_objective(factory, client):
     assert dataset.objective_key == objective.key
 
 
+@pytest.mark.parametrize('asset_name', [
+    'objective',
+    'algo',
+    'aggregate_algo',
+    'composite_algo',
+])
+def test_metadata(factory, client, asset_name):
```
Comment (jmorel) [inline tests/test_network.py:143]: Since this test is trying out multiple metadata values, we could parametrize the metadata as well.

There could be another layer of parametrization to specify what the input metadata value, what the output should and which error should be raised.

Here is an example of combining parametrization:
```python
@pytest.mark.parametrize('y', (1, 2))
@pytest.mark.parametrize('x', ('a', 'b'))
def test_foo(x, y):
    assert True
```
This will be executed in the following order:

* a-1
* a-2
* b-1
* b-2

See also for how to handle exceptions SubstraFoundation/substra-backend#189

File: tests/test_network.py [hunk 21]
```diff
@@ -124,6 +128,102 @@ def test_add_objective(factory, client):
...
+    #'algo',
+    #'aggregate_algo',
+    #'composite_algo',
+])
+@pytest.mark.parametrize('metadata,metadata_output,metadata_error', [
+    ({'foo': 'bar'}, {'foo': 'bar'}, {'foo': {"bar": "foo"}}),
+    (None, {}, {too_long: "bar"}),
+    ({}, {}, {"foo": too_long}),
+])
+def test_metadata(factory, client, asset_name, metadata, metadata_output, metadata_error):
```
Comment (jmorel) [inline tests/test_network.py:143]: This test could be splitted in 2 tests: test_valid_metadata and test_invalid_metadata. The former one would use the asset_name / metadata / metadata_output as args, and the latter one would use just asset_name and metadata. 

In order to handle all types of bad metadata in the same way, you would have to set the bad metadata on the spec_dict directly.

File: tests/test_network.py [hunk 22]
```diff
@@ -124,6 +128,102 @@ def test_add_objective(factory, client):
...
+    ('traintuple', 'algo'),
+    ('composite_traintuple', 'composite_algo'),
+])
+@pytest.mark.parametrize('metadata,metadata_output,metadata_error', [
+    ({'foo': 'bar'}, {'foo': 'bar'}, {'foo': {"bar": "foo"}}),
+    (None, {}, {too_long: "bar"}),
+    ({}, {}, {"foo": too_long}),
+])
+def test_metadata_traintuple(factory, client, asset_name, algo_type, default_dataset,
+                             metadata, metadata_output, metadata_error):
```
Comment (jmorel) [inline tests/test_network.py:176]: As for test_metadata, this test could be split in 2.

File: tests/test_network.py [hunk 23]
```diff
@@ -124,6 +128,102 @@ def test_add_objective(factory, client):
...
+        dataset=default_dataset,
+        data_samples=default_dataset.train_data_sample_keys,
+    )
+    composite_traintuple = client.add_composite_traintuple(composite_traintuple_spec)
+
+    aggregate_algo_spec = factory.create_aggregate_algo()
+    aggregate_algo = client.add_aggregate_algo(aggregate_algo_spec)
+    spec = factory.create_aggregatetuple(
+        algo=aggregate_algo,
+        worker='MyOrg1MSP',
```
Comment (jmorel) [inline tests/test_network.py:218]: The only hardcoded values in tests should be the values you're actually testing. In this case, The worker is `MyOrg1MSP` because it's the value of client.node_id, so you should use `client.node_id` instead.

File: tests/test_network.py [hunk 24]
```diff
@@ -124,6 +128,102 @@ def test_add_objective(factory, client):
...
+        traintuples=[composite_traintuple],
+        metadata={"foo": "bar"}
```
Comment (jmorel) [inline tests/test_network.py:220]: The test should be parametrized to test good values and bad values.

File: tests/test_network.py [hunk 25]
```diff
@@ -124,6 +128,102 @@ def test_add_objective(factory, client):
     assert dataset.objective_key == objective.key
 
 
+@pytest.mark.parametrize('asset_name', [
+    'dataset',
+    #'objective',
+    #'algo',
+    #'aggregate_algo',
+    #'composite_algo',
+])
```
Comment (jmorel) [inline tests/test_network.py:137]: It looks like you commented the wrong lines here :)

File: substratest/factory.py [hunk 26]
```diff
@@ -179,6 +179,7 @@ class DatasetSpec(_Spec):
     data_opener: str
     type: str
     description: str
+    metadata: typing.Dict[str, str] = None
```
Comment (samlesu) [inline substratest/factory.py:182]: We could do the same and add an another `_Metadata` type for this file.

File: tests/test_network.py [hunk 27]
```diff
@@ -124,6 +128,154 @@ def test_add_objective(factory, client):
...
+    'algo',
+    'aggregate_algo',
+    'composite_algo',
+])
+@pytest.mark.parametrize('metadata,metadata_output', [
+    ({'foo': 'bar'}, {'foo': 'bar'}),
+    (None, {}),
+    ({}, {}),
+])
+def test_metadata(factory, client, asset_name, metadata, metadata_output):
```
Comment (samlesu) [inline tests/test_network.py:143]: Could be named test_add_metadata to be consistent with other test names.

File: tests/test_network.py [hunk 28]
```diff
@@ -124,6 +128,154 @@ def test_add_objective(factory, client):
...
+    'aggregate_algo',
+    'composite_algo',
+])
+@pytest.mark.parametrize('metadata', [
+    'foo',
+    {'foo': {"bar": "foo"}},
+    {too_long: "bar"},
+    {"foo": too_long},
+])
+def test_invalid_metadata(factory, client, asset_name, metadata):
```
Comment (samlesu) [inline tests/test_network.py:166]: To be rename accordingly with the other test.

File: substratest/assets.py [hunk 29]
```diff
@@ -201,6 +201,7 @@ class Dataset(_Asset):
     name: str
     owner: str
     objective_key: str
+    metadata: typing.Dict[str, str]
```
[the remark this answers is not in the document]
Comment (maeldebon) [reply substratest/assets.py:204]: It doesn't work :/

[the remark this answers is not in the document]
Comment (samlesu) [reply substratest/assets.py:204]: Thanks for investigating. Let's keep the current version, this could be improved later.

File: substratest/factory.py (hunk 26 above)
Comment (samlesu) [reply substratest/factory.py:182]: Could be investigated later.

File: substratest/assets.py [hunk 30]
```diff
@@ -293,8 +297,9 @@ class Aggregatetuple(_Asset, _FutureMixin):
     compute_plan_id: str
     rank: int
     tag: str
+    metadata: typing.Dict[str, str]
     log: str
-    in_models: typing.List[InModel]
+    in_models: typing.List[InModel] = None
```
[the remark this answers is not in the document]
Comment (samlesu) [reply substratest/assets.py:302]: @maeldebon Could you answer this comment?

File: tests/test_network.py [hunk 31]
```diff
@@ -124,6 +125,106 @@ def test_add_objective(factory, client):
...
+@pytest.mark.parametrize('asset_name,algo_type', [
+    ('traintuple', 'algo'),
+    ('composite_traintuple', 'composite_algo'),
+])
+@pytest.mark.parametrize('metadata,metadata_output', [
+    ({'foo': 'bar'}, {'foo': 'bar'}),
+    (None, {}),
+    ({}, {}),
+])
+def test_traintuple_with_metadata(factory, client, asset_name, algo_type, default_dataset, metadata, metadata_output):
```
Comment (samlesu) [inline tests/test_network.py:180]: As said in a previous comment I think we should remove this test as well as the next one (`test_traintuple_with_invalid_metadata`).

In my opinion, the end to end tests should not test all the use cases and all the corner cases. This should be done in the unit test. So for maintainability and performance reasons, I think they are not needed.

File: substratest/factory.py [hunk 32]
```diff
@@ -329,7 +336,7 @@ class _BaseComputePlanSpec(_Spec, abc.ABC):
     aggregatetuples: typing.List[ComputePlanAggregatetupleSpec]
     testtuples: typing.List[ComputePlanTesttupleSpec]
 
-    def add_traintuple(self, algo, dataset, data_samples, in_models=None, tag=''):
+    def add_traintuple(self, algo, dataset, data_samples, in_models=None, tag='', metadata=None):
```
Comment (jmorel) [inline substratest/factory.py:339]: These `add_<asset>` are creating compute plan specs. It's been decided to handle compute plan specs in a separate PR, so they should be removed.

File: substratest/assets.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (maeldebon) [reply substratest/assets.py:302]: Yep, this asset is supposed to support being created without in_model keys, but it was never tested before. That's why I added this, to allow us to create it without these keys. 
But yes, this isn't directly related to the PR :/

File: substratest/factory.py (hunk 32 above)
Comment (maeldebon) [reply substratest/factory.py:339]: You're right, I'll do this!

**Comment by jmorel** [comment]
You need to rebase your branch apparently.

File: substratest/factory.py [hunk 33]
```diff
@@ -381,8 +388,9 @@ def add_composite_traintuple(self, composite_algo, dataset=None, data_samples=No
             train_data_sample_keys=_get_keys(data_samples),
             in_head_model_id=in_head_model.id if in_head_model else None,
             in_trunk_model_id=in_trunk_model.id if in_trunk_model else None,
-            out_trunk_model_permissions=out_trunk_model_permissions or DEFAULT_OUT_TRUNK_MODEL_PERMISSIONS,
+            out_trunk_model_permissions=out_trunk_model_permissions or DEFAULT_OUT_MODEL_PERMISSIONS,
```
Comment (samlesu) [inline substratest/factory.py:391]: What's the purpose of this change?
